File: frontend/src/services/api/cloudflare.js
// Cloudflare operations — zone settings (SSL/TLS, Speed, Caching, Security) and
// one-click hardening, layered on the existing Cloudflare DNS connection. Zones
// are addressed by their ServerKit DNS zone id (same as the /dns API).

export async function getCloudflareZoneSettings(zoneId) {
    return this.request(`/cloudflare/zones/${zoneId}/settings`);
}

export async function getCloudflareZoneSetting(zoneId, settingId) {
    return this.request(`/cloudflare/zones/${zoneId}/settings/${settingId}`);
}

export async function updateCloudflareZoneSetting(zoneId, settingId, value) {
    return this.request(`/cloudflare/zones/${zoneId}/settings/${settingId}`, {
        method: 'PATCH',
        body: { value },
    });
}

export async function applyCloudflareSettingsPreset(zoneId) {
    return this.request(`/cloudflare/zones/${zoneId}/settings/apply-preset`, {
        method: 'POST',
    });
}

// Purge the zone's Cloudflare cache. payload is { purge_everything: true } or
// { files: [...] } (and on Enterprise, hosts/prefixes/tags).
export async function purgeCloudflareCache(zoneId, payload) {
    return this.request(`/cloudflare/zones/${zoneId}/purge-cache`, {
        method: 'POST',
        body: payload,
    });
}

// WAF custom firewall rules (http_request_firewall_custom phase).
export async function getCloudflareWafRules(zoneId) {
    return this.request(`/cloudflare/zones/${zoneId}/waf/rules`);
}

export async function addCloudflareWafRule(zoneId, rule) {
    return this.request(`/cloudflare/zones/${zoneId}/waf/rules`, {
        method: 'POST',
        body: rule,
    });
}

export async function applyCloudflareWafPreset(zoneId, presetKey, params = {}) {
    return this.request(`/cloudflare/zones/${zoneId}/waf/presets/${presetKey}`, {
        method: 'POST',
        body: { params },
    });
}

export async function updateCloudflareWafRule(zoneId, rulesetId, ruleId, fields) {
    return this.request(`/cloudflare/zones/${zoneId}/waf/rulesets/${rulesetId}/rules/${ruleId}`, {
        method: 'PATCH',
        body: fields,
    });
}

export async function deleteCloudflareWafRule(zoneId, rulesetId, ruleId) {
    return this.request(`/cloudflare/zones/${zoneId}/waf/rulesets/${rulesetId}/rules/${ruleId}`, {
        method: 'DELETE',
    });
}

// Workers (edge hosting). Account is resolved from the zone server-side.
export async function getCloudflareWorkers(zoneId) {
    return this.request(`/cloudflare/zones/${zoneId}/workers`);
}

export async function deployCloudflareWorker(zoneId, payload) {
    return this.request(`/cloudflare/zones/${zoneId}/workers`, {
        method: 'POST',
        body: payload,
    });
}

export async function deleteCloudflareWorker(zoneId, name) {
    return this.request(`/cloudflare/zones/${zoneId}/workers/${encodeURIComponent(name)}`, {
        method: 'DELETE',
    });
}

export async function addCloudflareWorkerRoute(zoneId, pattern, script) {
    return this.request(`/cloudflare/zones/${zoneId}/workers/routes`, {
        method: 'POST',
        body: { pattern, script },
    });
}

export async function deleteCloudflareWorkerRoute(zoneId, routeId) {
    return this.request(`/cloudflare/zones/${zoneId}/workers/routes/${routeId}`, {
        method: 'DELETE',
    });
}

// Cloudflare Tunnels (cloudflared) — expose a local service through the edge.
export async function getCloudflareTunnels(zoneId) {
    return this.request(`/cloudflare/zones/${zoneId}/tunnels`);
}

export async function createCloudflareTunnel(zoneId, name) {
    return this.request(`/cloudflare/zones/${zoneId}/tunnels`, {
        method: 'POST',
        body: { name },
    });
}

export async function deleteCloudflareTunnel(zoneId, tunnelId) {
    return this.request(`/cloudflare/zones/${zoneId}/tunnels/${tunnelId}`, {
        method: 'DELETE',
    });
}

export async function getCloudflareTunnelInstall(zoneId, tunnelId) {
    return this.request(`/cloudflare/zones/${zoneId}/tunnels/${tunnelId}/install`);
}

export async function getCloudflareTunnelHostnames(zoneId, tunnelId) {
    return this.request(`/cloudflare/zones/${zoneId}/tunnels/${tunnelId}/hostnames`);
}

export async function addCloudflareTunnelHostname(zoneId, tunnelId, hostname, service) {
    return this.request(`/cloudflare/zones/${zoneId}/tunnels/${tunnelId}/hostnames`, {
        method: 'POST',
        body: { hostname, service },
    });
}

export async function removeCloudflareTunnelHostname(zoneId, tunnelId, hostname) {
    return this.request(`/cloudflare/zones/${zoneId}/tunnels/${tunnelId}/hostnames`, {
        method: 'DELETE',
        body: { hostname },
    });
}

// Developer platform — R2 buckets, KV namespaces, D1 databases (account-scoped).
export async function getCloudflareStorage(zoneId) {
    return this.request(`/cloudflare/zones/${zoneId}/storage`);
}

export async function createCloudflareR2Bucket(zoneId, name) {
    return this.request(`/cloudflare/zones/${zoneId}/storage/r2`, {
        method: 'POST',
        body: { name },
    });
}

export async function deleteCloudflareR2Bucket(zoneId, name) {
    return this.request(`/cloudflare/zones/${zoneId}/storage/r2/${encodeURIComponent(name)}`, {
        method: 'DELETE',
    });
}

export async function createCloudflareKvNamespace(zoneId, title) {
    return this.request(`/cloudflare/zones/${zoneId}/storage/kv`, {
        method: 'POST',
        body: { title },
    });
}

export async function deleteCloudflareKvNamespace(zoneId, namespaceId) {
    return this.request(`/cloudflare/zones/${zoneId}/storage/kv/${namespaceId}`, {
        method: 'DELETE',
    });
}

export async function createCloudflareD1Database(zoneId, name) {
    return this.request(`/cloudflare/zones/${zoneId}/storage/d1`, {
        method: 'POST',
        body: { name },
    });
}

export async function deleteCloudflareD1Database(zoneId, databaseId) {
    return this.request(`/cloudflare/zones/${zoneId}/storage/d1/${databaseId}`, {
        method: 'DELETE',
    });
}

File: builtin-extensions/serverkit-cloudflare-ops/backend/cloudflare.py
"""Cloudflare operations API — zone settings (SSL/TLS, Speed, Caching, Security)
and one-click hardening, on top of the existing Cloudflare DNS connection.

Zones are addressed by their ServerKit ``DNSZone`` id (same as the ``/dns`` API);
the service resolves the Cloudflare credential + zone id server-side. Reads are
available to any authenticated user; writes require admin. Every mutating call is
captured by the global audit fallback (method, route args, sanitized body).
"""
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity

from .cloudflare_service import CloudflareService, CloudflareError

cloudflare_bp = Blueprint('cloudflare', __name__)


def _require_admin():
    """Return the current user if admin, else ``None``."""
    from app.models.user import User
    user = User.query.get(get_jwt_identity())
    return user if (user and user.is_admin) else None


def _service_response(res):
    """Map a service result dict to JSON. A failed *provider* call (reached
    Cloudflare, got an error back) is a 502; resolution errors are raised as
    CloudflareError upstream and handled separately as 400s."""
    if not res.get('success'):
        return jsonify({'error': res.get('error', 'Cloudflare request failed')}), 502
    return jsonify(res)


# ── Zone settings ────────────────────────────────────────────────────────────

@cloudflare_bp.route('/zones/<int:zone_id>/settings', methods=['GET'])
@jwt_required()
def get_zone_settings(zone_id):
    try:
        res = CloudflareService.get_settings(zone_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/settings/apply-preset', methods=['POST'])
@jwt_required()
def apply_preset(zone_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    try:
        res = CloudflareService.apply_recommended(zone_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    # A preset is best-effort across many toggles; return 200 with the per-setting
    # report even when the plan gated some (the UI surfaces partial success).
    return jsonify(res)


@cloudflare_bp.route('/zones/<int:zone_id>/settings/<setting_id>', methods=['GET'])
@jwt_required()
def get_zone_setting(zone_id, setting_id):
    try:
        res = CloudflareService.get_setting(zone_id, setting_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/settings/<setting_id>', methods=['PATCH'])
@jwt_required()
def update_zone_setting(zone_id, setting_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    data = request.get_json(silent=True) or {}
    if 'value' not in data:
        return jsonify({'error': 'A "value" field is required'}), 400
    try:
        res = CloudflareService.update_setting(zone_id, setting_id, data['value'])
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


# ── Cache ────────────────────────────────────────────────────────────────────

@cloudflare_bp.route('/zones/<int:zone_id>/purge-cache', methods=['POST'])
@jwt_required()
def purge_cache(zone_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    data = request.get_json(silent=True) or {}
    try:
        res = CloudflareService.purge_cache(
            zone_id,
            everything=bool(data.get('purge_everything') or data.get('everything')),
            files=data.get('files'), hosts=data.get('hosts'),
            prefixes=data.get('prefixes'), tags=data.get('tags'))
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


# ── WAF custom rules ─────────────────────────────────────────────────────────

@cloudflare_bp.route('/zones/<int:zone_id>/waf/rules', methods=['GET'])
@jwt_required()
def list_waf_rules(zone_id):
    try:
        res = CloudflareService.list_waf_rules(zone_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/waf/rules', methods=['POST'])
@jwt_required()
def add_waf_rule(zone_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    data = request.get_json(silent=True) or {}
    try:
        res = CloudflareService.add_waf_rule(
            zone_id,
            description=data.get('description'),
            expression=data.get('expression'),
            action=data.get('action'),
            enabled=data.get('enabled', True))
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/waf/presets/<preset_key>', methods=['POST'])
@jwt_required()
def apply_waf_preset(zone_id, preset_key):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    data = request.get_json(silent=True) or {}
    try:
        res = CloudflareService.apply_waf_preset(zone_id, preset_key, data.get('params') or {})
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/waf/rulesets/<ruleset_id>/rules/<rule_id>',
                     methods=['PATCH'])
@jwt_required()
def update_waf_rule(zone_id, ruleset_id, rule_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    data = request.get_json(silent=True) or {}
    try:
        res = CloudflareService.update_waf_rule(zone_id, ruleset_id, rule_id, data)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/waf/rulesets/<ruleset_id>/rules/<rule_id>',
                     methods=['DELETE'])
@jwt_required()
def delete_waf_rule(zone_id, ruleset_id, rule_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    try:
        res = CloudflareService.delete_waf_rule(zone_id, ruleset_id, rule_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


# ── Workers (edge hosting) ───────────────────────────────────────────────────

@cloudflare_bp.route('/zones/<int:zone_id>/workers', methods=['GET'])
@jwt_required()
def list_workers(zone_id):
    try:
        res = CloudflareService.list_workers(zone_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/workers', methods=['POST'])
@jwt_required()
def deploy_worker(zone_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    data = request.get_json(silent=True) or {}
    try:
        res = CloudflareService.deploy_worker(
            zone_id,
            name=data.get('name'),
            code=data.get('code'),
            compatibility_date=data.get('compatibility_date'),
            route_pattern=data.get('route_pattern'))
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/workers/routes', methods=['POST'])
@jwt_required()
def add_worker_route(zone_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    data = request.get_json(silent=True) or {}
    try:
        res = CloudflareService.add_worker_route(zone_id, data.get('pattern'), data.get('script'))
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/workers/routes/<route_id>', methods=['DELETE'])
@jwt_required()
def delete_worker_route(zone_id, route_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    try:
        res = CloudflareService.delete_worker_route(zone_id, route_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


# Note: this dynamic <name> route is registered last so it can't shadow the more
# specific /workers/routes paths above.
@cloudflare_bp.route('/zones/<int:zone_id>/workers/<name>', methods=['DELETE'])
@jwt_required()
def delete_worker(zone_id, name):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    try:
        res = CloudflareService.delete_worker(zone_id, name)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


# ── Tunnels (cloudflared) ────────────────────────────────────────────────────

@cloudflare_bp.route('/zones/<int:zone_id>/tunnels', methods=['GET'])
@jwt_required()
def list_tunnels(zone_id):
    try:
        res = CloudflareService.list_tunnels(zone_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/tunnels', methods=['POST'])
@jwt_required()
def create_tunnel(zone_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    data = request.get_json(silent=True) or {}
    try:
        res = CloudflareService.create_tunnel(zone_id, data.get('name'))
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/tunnels/<tunnel_id>/install', methods=['GET'])
@jwt_required()
def tunnel_install(zone_id, tunnel_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    try:
        res = CloudflareService.get_tunnel_install(zone_id, tunnel_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/tunnels/<tunnel_id>/hostnames', methods=['GET'])
@jwt_required()
def tunnel_hostnames(zone_id, tunnel_id):
    try:
        res = CloudflareService.get_tunnel_hostnames(zone_id, tunnel_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/tunnels/<tunnel_id>/hostnames', methods=['POST'])
@jwt_required()
def add_tunnel_hostname(zone_id, tunnel_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    data = request.get_json(silent=True) or {}
    try:
        res = CloudflareService.add_tunnel_hostname(
            zone_id, tunnel_id, data.get('hostname'), data.get('service'))
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/tunnels/<tunnel_id>/hostnames', methods=['DELETE'])
@jwt_required()
def remove_tunnel_hostname(zone_id, tunnel_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    data = request.get_json(silent=True) or {}
    try:
        res = CloudflareService.remove_tunnel_hostname(zone_id, tunnel_id, data.get('hostname'))
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/tunnels/<tunnel_id>', methods=['DELETE'])
@jwt_required()
def delete_tunnel(zone_id, tunnel_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    try:
        res = CloudflareService.delete_tunnel(zone_id, tunnel_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


# ── Developer platform: R2 / KV / D1 ─────────────────────────────────────────

@cloudflare_bp.route('/zones/<int:zone_id>/storage', methods=['GET'])
@jwt_required()
def list_storage(zone_id):
    try:
        res = CloudflareService.list_storage(zone_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/storage/r2', methods=['POST'])
@jwt_required()
def create_r2_bucket(zone_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    data = request.get_json(silent=True) or {}
    try:
        res = CloudflareService.create_r2_bucket(zone_id, data.get('name'))
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/storage/r2/<bucket>', methods=['DELETE'])
@jwt_required()
def delete_r2_bucket(zone_id, bucket):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    try:
        res = CloudflareService.delete_r2_bucket(zone_id, bucket)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/storage/kv', methods=['POST'])
@jwt_required()
def create_kv_namespace(zone_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    data = request.get_json(silent=True) or {}
    try:
        res = CloudflareService.create_kv_namespace(zone_id, data.get('title'))
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/storage/kv/<namespace_id>', methods=['DELETE'])
@jwt_required()
def delete_kv_namespace(zone_id, namespace_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    try:
        res = CloudflareService.delete_kv_namespace(zone_id, namespace_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/storage/d1', methods=['POST'])
@jwt_required()
def create_d1_database(zone_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    data = request.get_json(silent=True) or {}
    try:
        res = CloudflareService.create_d1_database(zone_id, data.get('name'))
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)


@cloudflare_bp.route('/zones/<int:zone_id>/storage/d1/<database_id>', methods=['DELETE'])
@jwt_required()
def delete_d1_database(zone_id, database_id):
    if not _require_admin():
        return jsonify({'error': 'Admin access required'}), 403
    try:
        res = CloudflareService.delete_d1_database(zone_id, database_id)
    except CloudflareError as e:
        return jsonify({'error': str(e)}), 400
    return _service_response(res)
